pinot-controller/src/main/java/com/linkedin/pinot/controller/api/resources/PinotSegmentUploadRestletResource.java

```diff
@@ -291,26 +291,28 @@ private SuccessResponse uploadSegment(FormDataMultiPart multiPart, boolean enabl
 
       switch (uploadType) {
         case URI:
-          if (storageURI == null || storageURI.isEmpty()) {
+          if (currentSegmentLocationURI == null || currentSegmentLocationURI.isEmpty()) {
             throw new ControllerApplicationException(LOGGER, "Failed to get downloadURI, needed for URI upload",
                 Response.Status.BAD_REQUEST);
           }
           segmentMetadata =
-              getSegmentMetadata(crypterClassHeader, storageURI, tempEncryptedFile, tempDecryptedFile, tempSegmentDir,
+              getSegmentMetadata(crypterClassHeader, currentSegmentLocationURI, tempEncryptedFile, tempDecryptedFile, tempSegmentDir,
                   metadataProviderClass);
-          if (new URI(storageURI).getScheme().equals(CommonConstants.Segment.LOCAL_SEGMENT_SCHEME)) {
-            downloadURI = ControllerConf.constructDownloadUrl(segmentMetadata.getTableName(), segmentMetadata.getName(), provider.getVip());
+          if (new URI(currentSegmentLocationURI).getScheme().equals(CommonConstants.Segment.LOCAL_SEGMENT_SCHEME)) {
+            zkDownloadUri = ControllerConf.constructDownloadUrl(segmentMetadata.getTableName(), segmentMetadata.getName(), provider.getVip());
           } else {
-            downloadURI = storageURI;
+            LOGGER.info("Using configured data dir {}", _controllerConf.getDataDir());
```

Comment:
Why we need to do this? This is a backward-incompatible change and breaks the our production pipeline.
When we upload a segment with uri type, I need to save zkDownloadUri same as currentSegmentLocationURI.

Reply:
I think the intent here is for storageUri to be something purely under the control of controller, as the currentSegmentUri could be a path that controller might not have any control over (eg write permissions, segments of same table could in different locations, etc). Is this failing because datadir is not set? Or are you trying to avoid additional copy? One option could be an optimization in the code that infers that the current segment is already in the datadir and hence no move required.

Reply:
Yes, the major problem for us is that we don't have NAS mounting to all the controllers/servers for sharing segments.  All the segments are saved remotely on HDFS.

That's why our upload segment request will only contain segment download uri. We also expect the server to directly download segments from HDFS directly. Also for now we don't expect controller to save those data to HDFS, major concern is our HDFS quota.